dot/parachain/overseer/overseer.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package overseer
+
+import (
+	"fmt"
+	"github.com/ChainSafe/gossamer/internal/log"
+	"sync"
+)
+
+const COMMS_BUFFER_SIZE = 5
+
+var (
+	logger = log.NewFromGlobal(log.AddContext("pkg", "parachain-overseer"))
+)
+
+type Overseer struct {
+	wg     sync.WaitGroup
+	doneCh chan struct{}
+	stopCh chan struct{}
+
+	subsystems        map[Subsystem]*Context
+	subsystemMessages map[Subsystem]chan any
+}
+
+type ExampleSender struct {
+	senderChan chan any
+}
+
+func (s *ExampleSender) SendMessage(msg any) error {
+	s.senderChan <- msg
+	return nil
+}
+
+func NewOverseer() *Overseer {
+	return &Overseer{
+		doneCh:            make(chan struct{}),
+		stopCh:            make(chan struct{}),
+		subsystems:        make(map[Subsystem]*Context),
+		subsystemMessages: make(map[Subsystem]chan any),
+	}
+}
+
+func (o *Overseer) RegisterSubsystem(subsystem Subsystem) {
+	// TODO: determine best buffer size
+	comChan := make(chan any, COMMS_BUFFER_SIZE)
+	o.subsystemMessages[subsystem] = comChan
+	receiverChan := make(chan any, COMMS_BUFFER_SIZE)
+	o.subsystems[subsystem] = &Context{
+		Sender:   &ExampleSender{senderChan: comChan},
+		Receiver: receiverChan,
+	}
+}
+
+func (o *Overseer) Start() {
+	// start subsystems
+	for subsystem, context := range o.subsystems {
+		o.wg.Add(1)
+		go func(sub Subsystem, ctx *Context, wg sync.WaitGroup) {
+			defer wg.Done()
+			select {
+			case <-o.stopCh:
+				return
+			}
```

Comment:
you should put `default:` otherwise it will block until `stopCh` is triggered 
```suggestion
			select {
			case <-o.stopCh:
				return
			default:
			}
```

Reply:
Good point, I also moved this logic to the go routine inside the subsystem, instead of in the go routine that starts the subsystems.